taroscript/send.go

```diff
@@ -354,6 +359,132 @@ func CreateOutputCommitments(inputCommitment *commitment.TaroCommitment,
+// valid.
+func AreValidAnchorOutputIndexes(outputs []*taropsbt.VOutput) (bool, error) {
+	// Sanity check the output indexes provided by the sender. There must be
+	// at least two indexes; one for the receiver, and one for the change
+	// commitment for the sender.
+	if len(outputs) < 2 {
+		return false, ErrInvalidOutputIndexes
+	}
+
+	// If the indexes start from 0 and form a continuous range, then the
+	// resulting TX would be valid without any changes (Taro-only spend).
+	taroOnlySpend := true
+	sortedCopy := slices.Clone(outputs)
+	sort.Slice(sortedCopy, func(i, j int) bool {
+		return sortedCopy[i].AnchorOutputIndex <
+			sortedCopy[j].AnchorOutputIndex
+	})
+	for i := 0; i < len(sortedCopy); i++ {
+		if sortedCopy[i].AnchorOutputIndex != uint32(i) {
+			taroOnlySpend = false
+			break
+		}
+	}
+
+	return taroOnlySpend, nil
+}
+
+// CreateAnchorTx creates a template BTC anchor TX with dummy outputs.
+func CreateAnchorTx(outputs []*taropsbt.VOutput) (*psbt.Packet, error) {
+	// Check if our outputs are valid, and if we will need to add extra
+	// outputs to fill in the gaps between outputs.
+	taroOnlySpend, err := AreValidAnchorOutputIndexes(outputs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate the number of outputs we need for our template TX.
+	maxOutputIndex := uint32(len(outputs))
+
+	// If there is a gap in our outputs, we need to find the
+	// largest output index to properly size our template TX.
+	if !taroOnlySpend {
+		maxOutputIndex = 0
+		for _, out := range outputs {
+			if out.AnchorOutputIndex > maxOutputIndex {
+				maxOutputIndex = out.AnchorOutputIndex
+			}
+		}
+
+		// Output indexes are 0-indexed, so we need to increment this
+		// to account for the 0th output.
+		maxOutputIndex++
+	}
+
+	txTemplate := wire.NewMsgTx(2)
+	for i := uint32(0); i < maxOutputIndex; i++ {
+		txTemplate.AddTxOut(createDummyOutput())
+	}
+
+	spendPkt, err := psbt.NewFromUnsignedTx(txTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("unable to make psbt packet: %w", err)
+	}
+
+	return spendPkt, nil
+}
+
+// UpdateTaprootOutputKeys updates a PSBT with outputs embedding TaroCommitments
+// involved in an asset send. The sender must attach the Bitcoin input holding
+// the corresponding Taro input asset to this PSBT before finalizing the TX.
+// Locators MUST be checked beforehand.
+func UpdateTaprootOutputKeys(btcPacket *psbt.Packet,
+	outputs []*taropsbt.VOutput,
+	outputCommitments []*commitment.TaroCommitment) error {
+
+	// Add the commitment outputs to the BTC level PSBT now.
+	for idx := range outputs {
+		vOut := outputs[idx]
+		outputCommitment := outputCommitments[idx]
+
+		// The external output index cannot be out of bounds of the
+		// actual TX outputs. This should be checked earlier and is just
+		// a final safeguard here.
+		if vOut.AnchorOutputIndex >= uint32(len(btcPacket.Outputs)) {
+			return ErrInvalidOutputIndexes
+		}
+
+		btcOut := btcPacket.Outputs[vOut.AnchorOutputIndex]
```

Comment:
What about instances wherein you have multiple virtual outputs in a single Bitcoin output? Example would be: sending two collectibles to someone in the same transaction.

Reply:
Yes, this will be done at a later stage (meaning, another PR). See TODO about merging multiple Taro level commitments into a single one.